src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java

```diff
@@ -323,6 +324,14 @@ public static boolean shouldRetry(Throwable throwable, boolean isBatch) {
     return shouldRetry(throwable.getCause(), isBatch);
   }
 
+  public static String formatNestedExceptions(Throwable throwable, String message) {
+    if (throwable == null) {
+      return message;
+    }
+    message += "Caused by: " + throwable.getMessage() + " ";
+    return formatNestedExceptions(throwable.getCause(), message);
+  }
```

Comment:
In this case I think iteration might be clearer than recursion. You can use `Stream.iterate()` to write this
```suggestion
  public static String formatNestedExceptions(Throwable throwable) {
    return Stream.iterate(throwable, Objects::nonNull, Throwable::getCause)
        .map(t -> "Caused by: " + t.getMessage() + " ")
        .collect(Collectors.joining());
  }
```

Note that in some cases the exception type itself is an important part of the message so you may want to include it in the message string. Using `"Caused by: " + throwable + " "` would do this.

Reply:
Thanks for the suggestion! I left it as `t.getMessage()` since that included more detailed information about the error (e.g. which path already exists) that isn't included when I used `t`

Reply:
>  t.getMessage() since that included more detailed information about the error (e.g. which path already exists) that isn't included when I used t

That's odd since `Throwable.toString()` includes `getMessage()` in its result. Maybe this throwable overrides `toString()` to omit the message?

```
public class Throwable implements Serializable {
[...]
    public String toString() {
        String s = getClass().getName();
        String message = getLocalizedMessage();
        return (message != null) ? (s + ": " + message) : s;
    }
```